eclipse-theia/theia
# defer plugin contributions (preferences and tasks)

#### What it does


- fix #7673: with many VS Code extensions repetitive events are happening like update editor configurations and task schemas. This PR makes sure that such event defer changes in the same tick and perform them together.
- fix #7672: simplify `$fetchTasks` by removing filtering duplicate tasks. Testing showed that VS Code does not do it (see https://github.com/eclipse-theia/theia/issues/7672#issuecomment-620464015) and even after removing our implementation fails to return duplicate tasks. That's apparently an issue how the task extension treats the task identity: https://github.com/eclipse-theia/theia/issues/7681#issuecomment-620482611

TODO:
- [x] fix preference tests to check that there is not changes if schema set and unset in the same tick and there are if in different ticks
- [x] make sure that update/replace configuration calls from the extension host process are not resolved before preference changes are propagated to the extension host

#### How to test


- check out https://github.com/akosyakov/jupyterlab/tree/ak/fetch_tasks
```bash
git clone https://github.com/akosyakov/jupyterlab.git \
cd jupyterlab/ \
git checkout ak/fetch_tasks
```
- start Theia for [jlab.theia-workspace](https://github.com/akosyakov/jupyterlab/blob/ak/fetch_tasks/jlab.theia-workspace) workspace
```bash
yarn start ../../../jupyterlab/jlab.theia-workspace
```
- open README.md as an editor and for preview
- enable CPU profiler and reload the page
- check that that registering plugin contributions is fast, not like here https://github.com/eclipse-theia/theia/issues/7673#issuecomment-620007593

Testing task fetching:
- install https://github.com/akosyakov/vscode-fetch-tasks-test
```bash
cd plugins/ \
git clone https://github.com/akosyakov/vscode-fetch-tasks-test.git \
cd vscode-fetch-tasks-test/ \
yarn && yarn compile
```
- restart Theia to load the plugin
- use `Hello World` to see all found tasks in output view
- you can also test according to https://github.com/eclipse-theia/theia/pull/6058 and https://github.com/eclipse-theia/theia/pull/5975

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
